### PR TITLE
5150: Added sitemap of ting objects

### PR DESCRIPTION
**Pull request by arniloftsson**
#### Link to issue

Please add a link to the issue being addressed by this change.

#### Description

Please include a short description of the suggested change and the reasoning behind the approach you have chosen.

#### Screenshot of the result

If your change affects the user interface you should include a screenshot of the result with the pull request.

#### Checklist

- [ ] My complies with [our coding guidelines](../docs/code_guidelines.md).
- [ ] My code passes our static analysis suite. If not then I have added a comment explaining why this change should be exempt from the code standards and process.
- [ ] My code passes our continuous integration process. If not then I have added a comment explaining why this change should be exempt from the code standards and process.

#### Additional comments or questions

If you have any further comments or questions for the reviewer them please add them here.

### CODE REVIEW

File: modules/ding_seo/ding_seo.admin.inc [hunk 1]
```diff
@@ -16,12 +16,27 @@ function ding_seo_admin_settings($form, &$form_state) {
...
+
+  $form['sitemap_settings']['ding_seo_number_of_links_to_delete'] = [
+    '#type' => 'textfield',
+    '#title' => t('Number of links to delete'),
+    '#description' => t('How many links we delete to make number for new links. The links which have not been visited the longest are deleted'),
+    '#default_value' => variable_get('ding_seo_number_of_links_to_delete', 1000),
+  ];
 
-  return $form;
+  return system_settings_form($form);
```
Comment (holt83) [inline modules/ding_seo/ding_seo.admin.inc]: What about our `ding_seo_admin_settings_submit()` handler. Will this be called when using `system_settings_form()`?

File: modules/ding_seo/ding_seo.install [hunk 2]
```diff
@@ -12,3 +12,60 @@ function ding_seo_install() {
...
+        'type' => 'int',
+        'unsigned' => TRUE,
+        'not null' => TRUE,
+        'default' => 0,
+      ],
+    ],
+    'indexes' => [
+      'ding_seo_ting_object_id' => ['ting_object_id'],
+    ],
+    'primary key' => ['tid'],
```
Comment (holt83) [inline modules/ding_seo/ding_seo.install:56]: Could we use `ting_object_id` as primary key?

File: modules/ding_seo/ding_seo.module [hunk 3]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+ * Implements hook_cron().
+ *
+ * Updates the xmlsitemaptable with new links once a day
+ */
+function ding_seo_cron() {
+  $last_run = variable_get('ding_seo_last_run_sitemap', 0);
+
+  if (REQUEST_TIME > $last_run + 86400) {
+    ding_seo_update_sitemap();
+    watchdog('ding_seo', 'Updated sitemap: %time', array('%time' => REQUEST_TIME), WATCHDOG_INFO);
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: What if something went wrong? Perhaps this should be moved to `ding_seo_update_sitemap()`?

File: modules/ding_seo/ding_seo.module [hunk 4]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+
+/**
+ * Implements hook_cron().
+ *
+ * Updates the xmlsitemaptable with new links once a day
+ */
+function ding_seo_cron() {
+  $last_run = variable_get('ding_seo_last_run_sitemap', 0);
+
+  if (REQUEST_TIME > $last_run + 86400) {
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module:245]: Maybe use a constant for this.

File: modules/ding_seo/ding_seo.module [hunk 5]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+}
+
+/**
+ * Checks wether to put the ting_object into a sitemaps.
+ *
+ *  We only want ting_objects from the Library's own collection.
+ *  And we only want the ones the users have viewed as object or collection.
+ */
+function ding_seo_check_object_for_sitemap($ting_object, $view_mode) {
+  if ($view_mode != 'full' && $view_mode != 'collection_list') {
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: Nit: Think this check would be better placed in `ding_seo_ding_entity_view()` and then not passing `$view_mode`

File: modules/ding_seo/ding_seo.module [hunk 6]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+
+    if (count($objects) > 0) {
+      ding_seo_update_sitemap_table($objects);
+      variable_set('xmlsitemap_regenerate_needed', TRUE);
+      variable_set('ding_seo_sitemap_last_check', time());
+    }
+  } catch (Exception $exception) {
+    watchdog_exception('ding_seo', $exception);
+  }
+
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: Empty line

File: modules/ding_seo/ding_seo.module [hunk 7]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+}
+
+/**
+ * Get the new links from last update.
+ */
+function ding_seo_get_objects($last_check ) {
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: whitespace

File: modules/ding_seo/ding_seo.admin.inc [hunk 8]
```diff
@@ -16,12 +16,27 @@ function ding_seo_admin_settings($form, &$form_state) {
...
+    '#type' => 'fieldset',
+    '#title' => t('Sitemap settings'),
+    '#description' => t('Settings for processing Ting objects into sitemap links'),
+  ];
+
+  $form['sitemap_settings']['ding_seo_maximum_links_in_sitemap'] = [
+    '#type' => 'textfield',
+    '#title' => t('Maximum Number of links in sitemap'),
+    '#description' => t('The maximum number of links we allow in sitemaps. To large a number leads to performance strain on the server'),
+    '#default_value' => variable_get('ding_seo_maximum_links_in_sitemap', 50000),
```
Comment (holt83) [inline modules/ding_seo/ding_seo.admin.inc]: I would suggest to use a constant for this default and `ding_seo_number_of_links_to_delete`.

File: modules/ding_seo/ding_seo.module [hunk 9]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+ *
+ *  We only want ting_objects from the Library's own collection.
+ *  And we only want the ones the users have viewed as object or collection.
+ */
+function ding_seo_check_object_for_sitemap($ting_object, $view_mode) {
+  if ($view_mode != 'full' && $view_mode != 'collection_list') {
+    return;
+  }
+
+  if ($ting_object->getAc_source() == 'Bibliotekskatalog') {
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: Could we use `$entity->is('library_material')` to make this provider independent?

File: modules/ding_seo/ding_seo.module [hunk 10]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+ * Check if our ting_object_sitemap table exceedes the maximum number of link.
+ *
+ * We trim the links which haven't been visited for the longest time.
+ */
+function ding_seo_check_sitemap_size() {
+  $max_size = variable_get('ding_seo_maximum_links_in_sitemap', 50000);
+  $size = (int) db_query("SELECT count(*) FROM {ting_object_sitemap}")->fetchField() ?: 0;
+
+  if ($size > $max_size) {
+    $number_to_delete = variable_get('ding_seo_number_of_links_to_delete', 1000);
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module]: Have you considered ditching this setting and just deleting `$size - $max_size` (if positive) every time?

File: modules/ding_seo/ding_seo.module [hunk 11]
```diff
@@ -226,3 +228,185 @@ function ding_seo_get_person_nodes(array $names) {
...
+/**
+ * Insert the new links into xmlsitemaptable.
+ */
+function ding_seo_update_sitemap_table($objects) {
+  $maxid = (int) db_query("SELECT max(id) FROM {xmlsitemap} WHERE type = 'ting_object'")->fetchField() ?: 0;
+  $chunks = array_chunk($objects, 1000, TRUE);
+  foreach ($chunks as $chunk) {
+    // We could use xmlsitemap_link_save but it takes only one link at a time. For performance reasons
+    // we update the database table our selves.
+    $query = db_insert('xmlsitemap')->fields(array('id', 'type', 'loc', 'language', 'priority'));
```
Comment (holt83) [inline modules/ding_seo/ding_seo.module:310]: Have you considered just using xmlsitemap_link_save() directly everytime an object is viewed?

Thereby ditching our custom table all together.

I noticed that xmlsitemap rows has an `lastmod` timestamp which could be used for ordering and cleanup?

File: modules/ding_seo/ding_seo.install (hunk 2 above)
Comment (arniloftsson) [reply modules/ding_seo/ding_seo.install:56]: Yes we could. But I use the tid as a id in the xmlsitemaptable so we need it anyway

File: modules/ding_seo/ding_seo.module (hunk 4 above)
Comment (arniloftsson) [reply modules/ding_seo/ding_seo.module:245]: I think using constant makes the code less readable and is not done in other modules.

Comment (holt83) [reply modules/ding_seo/ding_seo.module:245]: If you use a constant like `DAY_IN_SECONDS` you can immediately see how loing the period is. I don't know how that could be less readable than having to calculate first :)

Alternatively a comment like `// Run once every day` would also make it more readable.

**Comment by holt83** [comment]
@arniloftsson 

Thanks for the updates 👍  

I took the liberty to make a [few small changes after merge](https://github.com/ding2/ding2/pull/1587/commits/006e9a42b6e79386d028bcd22cdeeb996a9ca283):
- Removed `$view_mode` parameter when calling `ding_seo_check_object_for_sitemap()`. It's not used anymore.
- Renamed constant to `DING_SEO_DEFAULT_MAXIMUM_SITEMAP_LINKS`

**Pull request merged**